### PR TITLE
22039 - MBR Filing Status

### PR DESCRIPTION
**Pull request by ArwenQin**
*Issue #:*
https://github.com/bcgov/entity/issues/22039

*Description of changes:*
- Updated the status and action displays for different filing intermediate states
- FE will display as  "Future Effective - Paid", action will be Open Application/Open Registration". No delete action.
- Updated for AWAITING_REVIEW, CHANGE_REQUESTED, REJECTED
- Added unit tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-auth license (Apache 2.0).

### CODE REVIEW

File: auth-web/src/stores/business.ts
Comment (ArwenQin) [inline auth-web/src/stores/business.ts:94]: I tried to get the status and effectiveDate property. But the Business Object is still not getting the state and effectiveDate property by doing so, can I get some guidance or suggestions?

**Comment by ArwenQin** [comment]
Will add unit tests later

File: auth-web/src/composables/affiliations-factory.ts
Comment (seeker25) [inline auth-web/src/composables/affiliations-factory.ts:86]: Would recommend moving this to the API if possible

File: auth-web/src/stores/business.ts
Comment (seeker25) [reply auth-web/src/stores/business.ts:94]: take a look at the search api?
image

**Comment by ArwenQin** [comment]
FE:
image
AWAITING_REVIEW:
image
CHANGE_REQUESTED:
image

File: auth-web/src/stores/business.ts
Comment (ArwenQin) [reply auth-web/src/stores/business.ts:94]: > I tried to get the status and effectiveDate property. But the Business Object is still not getting the state and effectiveDate property by doing so, can I get some guidance or suggestions?

effectiveDate property is working! I assume the state/status is not updated in the api?

Comment (severinbeauvais) [reply auth-web/src/stores/business.ts:94]: The Auth API probably needs a small update to pass things along. I assume @vysakh-menon-aot will work on it next as part of his ticket.

Comment (vysakh-menon-aot) [reply auth-web/src/stores/business.ts:94]: image

**Comment by severinbeauvais** [comment]
image

This doesn't match the latest UI design. To be honest, Andy is still working on them, but between the 3 of us we should be able to figure out what to do.

https://www.figma.com/design/wVF8EcUCUR20YyqA0wHVJp/Continuation-In-V5?node-id=2983-13864&t=tdBPWazBJSVFFHYW-0

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2943-r0ywaf4j.web.app

**Comment by ArwenQin** [comment]
> image
> 
> This doesn't match the latest UI design. To be honest, Andy is still working on them, but between the 3 of us we should be able to figure out what to do.
> 
> https://www.figma.com/design/wVF8EcUCUR20YyqA0wHVJp/Continuation-In-V5?node-id=2983-13864&t=tdBPWazBJSVFFHYW-0

I will update!

File: auth-web/src/composables/affiliations-factory.ts
Comment (severinbeauvais) [reply auth-web/src/composables/affiliations-factory.ts]: Arwen, did you do this because some combined states are more important than others and we want to show those?

Example 1: don't show future effective -- awaiting review is more important
Example 2: don't show future effective -- change requested is more important)?

If so, please make this clear in the code with sufficient comments.

File: auth-web/src/composables/affiliations-factory.ts
Comment (severinbeauvais) [inline auth-web/src/composables/affiliations-factory.ts:91]: I'm confused why you need lines 86-88. Won't like 90 handle it?

File: auth-web/src/stores/business.ts
Comment (ArwenQin) [inline auth-web/src/stores/business.ts:95]: I updated this:
if there is a state ("ACTIVE" or "HISTORICAL"), business.status = state
if there is a draftStatus, the business.status would be the draftStatus.
It should work, but after I update it, there is an error:
image
Could you please help debug? @severinbeauvais  Thanks!

File: auth-web/src/composables/affiliations-factory.ts
Comment (ArwenQin) [reply auth-web/src/composables/affiliations-factory.ts]: Yes, I will add comments

File: auth-web/src/stores/business.ts
Comment (severinbeauvais) [reply auth-web/src/stores/business.ts:95]: You have a null string on which `toUpperCase` is being called. Check your data. Use conditional chaining if needed.

File: auth-web/src/composables/affiliations-factory.ts
Comment (ArwenQin) [reply auth-web/src/composables/affiliations-factory.ts:91]: Agree, I will fix it.

**Comment by severinbeauvais** [comment]
The third item here should be in Awaiting Review status.

image

image

File: auth-web/src/composables/affiliations-factory.ts
Comment (severinbeauvais) [inline auth-web/src/composables/affiliations-factory.ts:114]: Is this code needed or will line 88 cover it?

**Comment by ArwenQin** [comment]
> The third item here should be in Awaiting Review status.
> 
> image
> 
> image

This is because the draftStatus is not correctly read yet, I will fix that issue!

File: auth-web/src/composables/affiliations-factory.ts
Comment (severinbeauvais) [reply auth-web/src/composables/affiliations-factory.ts]: Oh, sorry, I see you already have those comments further down (lines 92-93).

@seeker25 , what are you thinking should be in the API instead?

File: auth-web/src/composables/affiliations-factory.ts
Comment (severinbeauvais) [reply auth-web/src/composables/affiliations-factory.ts:91]: Also, isn't there always a `status` property in the business object? In this case, the following code will not be run...?

File: auth-web/src/stores/business.ts
Comment (ArwenQin) [reply auth-web/src/stores/business.ts:95]: Even when I set the status to be null, or empty string, there is no error. The error only shows when I set the status to be resp.draftStatus.

File: auth-web/src/composables/affiliations-factory.ts
Comment (seeker25) [reply auth-web/src/composables/affiliations-factory.ts]: return the enum that the UI needs and not create these pseudo statuses that live just in the UI, the result will be cleaner UI code that we could easily rewire up to a new UI if required

Comment (seeker25) [reply auth-web/src/composables/affiliations-factory.ts]: There's already way too many conditional branches in the UI for this kind of thing @hfekete stated this before

Comment (severinbeauvais) [reply auth-web/src/composables/affiliations-factory.ts]: @seeker25 , please wait for the next commit. I believe a lot of this code will resolve to a single statement.

By the way, this code is not about creating pseudo-statuses -- it's about changing an enum to a user-friendly string. Do  you agree that this belongs in the UI?

PS - If I'm not mistaken, this code is handling the response of the `affiliations` call.

Comment (seeker25) [reply auth-web/src/composables/affiliations-factory.ts]: Why not just return the correct status enum from the API, then map that to a UI user friendly string?

Comment (severinbeauvais) [reply auth-web/src/composables/affiliations-factory.ts]: > Why not just return the correct status enum from the API, then map that to a UI user friendly string? No conditions needed

I believe that is exactly what the code is doing presently. `BusinessState` is a string lookup enum:

image

Comment (seeker25) [reply auth-web/src/composables/affiliations-factory.ts]: Looks conditional to me and not just looking at the enum:

image

Comment (severinbeauvais) [reply auth-web/src/composables/affiliations-factory.ts]: This logic is being rewritten. Please wait for the next commit.

Comment (seeker25) [reply auth-web/src/composables/affiliations-factory.ts]: anyways just my thoughts, I've seen people try to refactor this code and it proved very difficult

File: auth-web/src/composables/affiliations-factory.ts
Comment (ArwenQin) [reply auth-web/src/composables/affiliations-factory.ts]: Updated the logic

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2943-r0ywaf4j.web.app

File: auth-web/src/composables/affiliations-factory.ts
Comment (ArwenQin) [reply auth-web/src/composables/affiliations-factory.ts:91]: 1.  When the business has a effectiveDate and no other draft statuses (like "AWAITING_REVIEW" and "CHANGE_REQUESTED"), the draftStatus fetched from api is PAID. So the business.status would be PAID. No need to check business.effectiveDate again.
2. When the business has a effectiveDate and has a draft status, the draftStatus (like "AWAITING_REVIEW") will override PAID. The business.status would be the draftStatus fetched from api, so PAID (Effective Effective) won't show.
3. Same for PENDING

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (severinbeauvais) [inline auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:147]: 1. A filing may be in Paid (or Approved) status temporarily if the Filer is stuck. I think you also need to check the future-effective flag.
2. A tooltip should be attached to this icon. Andy has prepared a short tooltip that includes the FE date.

Comment (ArwenQin) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:147]: Just to confirm:
- Paid and Future Effective: show Future Effective and the tooltip that includes the FE date
- Paid and not FE (the Filer is stuck): show Paid.

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (severinbeauvais) [inline auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:538]: I recommend using a class of "ml-1" for alignment.

Also, can you use a color of `#F8661A` ? If that's acceptable then I think you can specify it right in the template (color="xxx"), so you don't need this CSS.

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (severinbeauvais) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:147]: Correct!

**Comment by severinbeauvais** [comment]
These NRs have the wrong status:

image

**Comment by ArwenQin** [comment]
> These NRs have the wrong status:
> 
> image

May I know what is the correct status for these NRs. They are in "Pending Staff Review" currently, before my changes

**Comment by severinbeauvais** [comment]
> > These NRs have the wrong status:
> 
> May I know what is the correct status for these NRs. They are in "Pending Staff Review" currently, before my changes

My mistake. You are correct: these NRs are in Pending Staff Review status. Please ignore my comment.

**Review by seeker25 — APPROVED** [review]
LGTM, someone from entities should approve though as well

File: auth-web/src/util/constants.ts
Comment (severinbeauvais) [inline auth-web/src/util/constants.ts:414]: To be clear, this state could mean:
- a future effective filing
- an immediate filing that the Filer hasn't processed yet (maybe "Paid | Pending"? ask Andy) (this could also be an overdue future effective filing)

In order to keep the strings here, I think you'll need a new business state here (PAID_FE?), and then you'll need to select between them in affiliations-factory.ts.

Also, you should handle APPROVED state the same as PAID (and Approved could also be FE).

**Comment by severinbeauvais** [comment]
https://bcregistry-account-dev--pr-2943-r0ywaf4j.web.app/account/2288/business

image

image

image

It looks like `state` is null/undefined above.

**Comment by ArwenQin** [comment]
> https://bcregistry-account-dev--pr-2943-r0ywaf4j.web.app/account/2288/business
> 
> image
> 
> image
> 
> image
> 
> It looks like `state` is null/undefined above.
> 
> **Update: this may be because you're doing this:**
> 
> image
> 
> **but the status (in this case, APPROVED) in not in the enum.**

Yes, it's due to APPROVED, I will update now.

File: auth-web/src/util/constants.ts
Comment (severinbeauvais) [reply auth-web/src/util/constants.ts:414]: Sample future-effective, approved continuation application:

https://dev.business.bcregistry.gov.bc.ca/Tzg27dGHVo?filing_id=150492&accountid=2288

**Comment by severinbeauvais** [comment]
Please also update the app version again. (Your change was swallowed by the rebase.)

File: auth-web/src/util/constants.ts
Comment (severinbeauvais) [reply auth-web/src/util/constants.ts:414]: Would it make sense to have a "getBusinessState" function in which to encapsulate the string lookup and special logic? Maybe we don't need the enum either if everything is in one place.

Idea:
```
function getBusinessState (status): string {
   switch (status) {
      case 'PAID': return (isFutureEffective) ? 'Future Effective' : 'Paid | Pending'
      case 'DRAFT': return 'Draft'
      ...
      default: return '' // or 'Unknown'?
   }
}
```

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (ArwenQin) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:157]: Updated!
- Paid and Future Effective (approved or not-approved with no other status): show Future Effective and the tooltip that includes the FE date
- Paid and not FE (the Filer is stuck): show Paid | Pending.
- Approved: show Approved if it's not a FE

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (ArwenQin) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue]: Updated

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2943-r0ywaf4j.web.app

File: auth-web/src/util/constants.ts
Comment (ArwenQin) [reply auth-web/src/util/constants.ts]: Updated status:
- PAID (filer pending)
- APPROVED
- PAID_FE

I think may be better to keep the enums, because I used them in the AffliationAction file to determine actions for different statuses; and computed some statuses when fetch from api

**Comment by ArwenQin** [comment]
2 action items left for now:
- The 'Paid | Pending' status, for immediate filing, and for overdue FE (confirming with Andy)
- Will add unit tests

**Comment by severinbeauvais** [comment]
Re:
image

Ask Andy but I think we should display the future effective time also.

**Comment by severinbeauvais** [comment]
Looks great! (BCREG0020 / Severin Dev Account)

image

- Do we really want to say "Manage Business" for a FE continuation? It's not a business yet. I think it should saw "Open Application". Confirm with Andy.

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (severinbeauvais) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:157]: Perfect, thanks!

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (severinbeauvais) [inline auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:333]: Does this fail safely if dateString is null/undefined/empty?

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2943-r0ywaf4j.web.app

File: auth-web/src/util/constants.ts
Comment (severinbeauvais) [reply auth-web/src/util/constants.ts]: OK. I appreciate the effort to not make too many changes. Hopefully when this is ported to another UI, we can simplify the code somewhat.

File: auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue
Comment (ArwenQin) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue:333]: Will update this.

Comment (ArwenQin) [reply auth-web/src/components/auth/manage-business/AffiliatedEntityTable.vue]: Updated, and show the date + time

**Comment by ArwenQin** [comment]
Will update the package version when it's ready to merge, in case rebase is needed

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-account-dev--pr-2943-r0ywaf4j.web.app

**Review by severinbeauvais — APPROVED** [review]
LGTM!

Please go ahead and update the app version, and then we can merge this.

~While you're here, can you fix this bug? Should say "This continuation application" in this case.~ _will be new ticket_

**Comment by severinbeauvais** [comment]
Oh, here's another one:

image

Should be "This registration".

**Comment by severinbeauvais** [comment]
And another. If this is too much, let me know and I'll create a separate bug ticket.

image

**Pull request merged**
